drivers/gpio/gpio_cy8c95xx.c

```diff
@@ -0,0 +1,317 @@
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	uint8_t pin_data = 0;
+	int rc = 0;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(drv_data->lock, K_FOREVER);
+
+	rc = i2c_reg_read_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			       CY8C95XX_REG_INPUT_DATA0 + cfg->port_num, &pin_data);
+
+	if (rc == 0) {
+		*value = pin_data;
+	}
+
+	k_sem_give(drv_data->lock);
+	return rc;
+}
+
+static int port_write(const struct device *dev,
+		      gpio_port_pins_t mask,
+		      gpio_port_value_t value,
+		      gpio_port_value_t toggle)
+{
+	const struct cy8c95xx_config *cfg = dev->config;
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	uint8_t *outp = &drv_data->pin_state.data_out;
+	uint8_t out = ((*outp & ~mask) | (value & mask)) ^ toggle;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(drv_data->lock, K_FOREVER);
+
+	int rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    CY8C95XX_REG_OUTPUT_DATA0 + cfg->port_num, out);
+
+	if (rc == 0) {
+		*outp = out;
+	}
+	k_sem_give(drv_data->lock);
+
+	LOG_DBG("write msk %04x val %04x tog %04x => %04x: %d",
+		mask, value, toggle, out, rc);
+
+	return rc;
+}
+
+static int port_set_masked(const struct device *dev,
+			   gpio_port_pins_t mask,
+			   gpio_port_value_t value)
+{
+	return port_write(dev, mask, value, 0);
+}
+
+static int port_set_bits(const struct device *dev,
+			 gpio_port_pins_t pins)
+{
+	return port_write(dev, pins, pins, 0);
+}
+
+static int port_clear_bits(const struct device *dev,
+			   gpio_port_pins_t pins)
+{
+	return port_write(dev, pins, 0, 0);
+}
+
+static int port_toggle_bits(const struct device *dev,
+			    gpio_port_pins_t pins)
+{
+	return port_write(dev, 0, 0, pins);
+}
+
+static int pin_interrupt_configure(const struct device *dev,
+				   gpio_pin_t pin,
+				   enum gpio_int_mode mode,
+				   enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+/**
+ * @brief Initialization function of CY8C95XX
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int cy8c95xx_init(const struct device *dev)
+{
+	const struct cy8c95xx_config *cfg = dev->config;
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	int rc;
+	uint8_t data = 0;
+
+	k_sem_take(drv_data->lock, K_FOREVER);
+
+	drv_data->i2c_master = device_get_binding(cfg->i2c_master_dev_name);
+	if (!drv_data->i2c_master) {
+		LOG_ERR("%s: no bus %s", dev->name, cfg->i2c_master_dev_name);
+		rc = -EINVAL;
+		goto out;
+	}
+
+	i2c_reg_read_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+			  CY8C95XX_REG_ID, &data);
+	LOG_DBG("cy8c95xx device ID %02X", data & 0xF0);
+	if ((data & 0xF0) != 0x20) {
+		LOG_WRN("driver only support [0-2] ports operations");
+	}
+
+	/* Reset state mediated by initial configuration */
+	drv_data->pin_state = (struct cy8c95xx_pin_state) {
+		.dir = 0xFF,
+		.data_out = 0x00,
```

Comment:
We also need to keep `data_out = 0xFF`, since that's the default power up value. The `write_pin_state` writes to the data output register first, pin direction register later. With `data_out = 0x00` we would observe glitch on the outputs during device init.